consensus/src/aggregation/engine.rs

```diff
@@ -0,0 +1,799 @@
+    }
+
+    /// Runs the engine until the context is stopped.
+    ///
+    /// The engine will handle:
+    /// - Requesting and processing digests from the automaton
+    /// - Timeouts
+    ///   - Refreshing the Epoch
+    ///   - Rebroadcasting Acks
+    /// - Messages from the network:
+    ///   - Acks from other validators
+    pub fn start(mut self, network: (NetS, NetR)) -> Handle<()> {
+        self.context.spawn_ref()(self.run(network))
+    }
+
+    /// Inner run loop called by `start`.
+    async fn run(mut self, (net_sender, net_receiver): (NetS, NetR)) {
+        let (mut net_sender, mut net_receiver) = wrap((), net_sender, net_receiver);
+        let mut shutdown = self.context.stopped();
+
+        // Initialize the epoch
+        let (latest, mut epoch_updates) = self.monitor.subscribe().await;
+        self.epoch = latest;
+
+        // Initialize Journal
+        let journal_cfg = JConfig {
+            partition: self.partition.clone(),
+            compression: self.journal_compression,
+            codec_config: (),
+            write_buffer: self.journal_write_buffer,
+        };
+        let journal = Journal::init(self.context.with_label("journal"), journal_cfg)
+            .await
+            .expect("init failed");
+        self.replay(&journal).await;
+        self.journal = Some(journal);
+
+        // Initialize the tip manager
+        self.safe_tip.init(
+            self.validators
+                .participants(self.epoch)
+                .expect("unknown participants"),
+        );
+
+        loop {
+            self.metrics.tip.set(self.tip as i64);
+
+            // Propose a new digest if we are processing less than the window
+            let next = self.next();
+            if next < self.tip + self.window {
+                trace!("requesting new digest: index {}", next);
+                self.get_digest(next);
+                continue;
+            }
+
+            // Get the rebroadcast deadline for the next index
+            let rebroadcast = match self.rebroadcast_deadlines.peek() {
+                Some((_, &deadline)) => Either::Left(self.context.sleep_until(deadline)),
+                None => Either::Right(future::pending()),
+            };
+
+            // Process the next event
+            select! {
+                // Handle shutdown signal
+                _ = &mut shutdown => {
+                    debug!("shutdown");
+                    break;
+                },
+
+                // Handle refresh epoch deadline
+                epoch = epoch_updates.next() => {
+                    // Error handling
+                    let Some(epoch) = epoch else {
+                        error!("epoch subscription failed");
+                        break;
+                    };
+
+                    // Refresh the epoch
+                    debug!(current=self.epoch, new=epoch, "refresh epoch");
+                    assert!(epoch >= self.epoch);
+                    self.epoch = epoch;
```

Comment:
Unlike `ordered-broadcast`, I suspect `epoch` will be rigidly associated with a range of indices.

We "should" be ok gathering signatures from the current epoch on any historical data (and need to for liveness) but just calling out as a point of discussion.